lib/ignored-paths.js

```diff
@@ -190,34 +190,46 @@ IgnoredPaths.prototype.contains = function(filepath, category) {
 
 /**
  * Returns a list of dir patterns for glob to ignore
- * @returns {string[]} list of glob ignore patterns
+ * @returns {function()} method to check whether a folder should be ignored by glob.
  */
-IgnoredPaths.prototype.getIgnoredFoldersGlobPatterns = function() {
-    let dirs = DEFAULT_IGNORE_DIRS;
+IgnoredPaths.prototype.getIgnoredFoldersGlobChecker = function() {
 
-    if (this.options.ignore) {
+    /**
+     * Adds `"*"` at the end of `"node_modules/"`,
+     * so that subtle directories could be re-included by .gitignore patterns
+     * such as `"!node_modules/should_not_ignored"`
+     */
+    const dirs = DEFAULT_IGNORE_DIRS.map(function(dir) {
```

Comment:
It seems my previous comment https://github.com/eslint/eslint/pull/6783#discussion_r73985714 was lost in the rebase.  I still think it would be cleaner to just make the `DEFAULT_IGNORE_DIRS` the way we want them.